### ISSUE TITLE
tesseract.exe still not flushing stdout

### ISSUE THREAD

**Issue by dmigowski** [tesseract-ocr/tesseract#3085]
I am using the tesseract build from https://github.com/UB-Mannheim/tesseract/wiki, which was compiled on 2020-03-28, but issue #2859 does not seem to be fixed for me. 

I also call Tesseract from Java with the following command line:

`tesseract.exe stdin out_overlay.pdf --dpi 300 -c textonly_pdf=1 -c stream_filelist=true -l deu pdf`

However I only receive any text on stdout when the process ends. I believe the patch in #2859 doesn't work in my case. 

I would love some option to make stdout unbuffered completely, which is ideal for me. Also the patch seems to do output only AFTER finishing a page and for better interleaving of image generation with the tesseract process I need a flush directly after the "Page n : <filename>" output. 

Thanks alot if anyone could implement this (and maybe even provide a Windows build for me, which I tried but didn't get to run after a few hours, else I would have implemented this for myself).

**Comment by dmigowski** [comment]
Just managed to build Tesseract for myself on Windows with VC and with that version it works, maybe because it doesn't use GLIBC. I don't need a fix here anymore, but it would be nice to have it so the Uni Mannheim builds will work again.

**Comment by stweil** [comment]
Wouldn't it be easier to use the Tesseract API directly? Then you are free to do buffering as you like.

The build process which is used by UB Mannheim is also used with [GitHub actions](https://github.com/UB-Mannheim/tesseract/actions?query=workflow%3A%22Cross+build+for+Windows%22). All relevant scripts are available [online](https://github.com/UB-Mannheim/tesseract/tree/windows/.github/workflows) and can also be used locally on Debian, Ubuntu and compatible Linux distributions.

**Comment by stweil** [comment]
@dmigowski, at which point of the output do you expect (currently missing) flushing? Completely unbuffered output is very inefficient and usually not a good solution.

**Comment by dmigowski** [comment]
@stweil, I am writing a Java solution and using the API directly requires me to build a JNI layer which is not impossible but also unnessecary hard because the main work is not done by the IO but within the Tesseract process itself. 

Also I found that the most efficient way to use Tesseract ist instantiating a number of Tesseract processes equal to the number of CPU cores, and then piping the names of the page images to the processes. Tesseracts internal multi cpu handling isn't that sofisticated.

To achive this I require Tesseract to flush the line starting with "Page " and ending with the filename. I use this marker to start extracting a new image from the source PDF and pushing it to Tess. 

Also while unbuffered IO is slow, the few characters printed to the command line from tesseract are surely negligable compared to the other work done in the process, but as said I need only a flush after the beginning of each page. I believe the flush should be in baseapi.cpp around line 1021:

```
    if (pix == nullptr) {
      tprintf("Image file %s cannot be read!\n", pagename);
      return false;
    }
    tprintf("Page %d : %s\n", page, pagename);     <- Flush after this!!
    bool r = ProcessPage(pix, page, pagename, retry_config,
                         timeout_millisec, renderer);
    pixDestroy(&pix);
    if (!r) return false;
```

The strange this here is that tprintf seems to write to stderr which should be unbuffered (at least #2859 says it) but there either is some redirection in place or some other fuckery I don't understand.

**Comment by dmigowski** [comment]
And while you are on it, a flush after line 1058 should help users that want to use Tesseract single threaded but want to display some kind of progress bar:

```
      pix = (data) ? pixReadMemFromMultipageTiff(data, size, &offset)
                   : pixReadFromMultipageTiff(filename, &offset);
    }
    if (pix == nullptr) break;
    tprintf("Page %d\n", page + 1);  <-- Flush here
    char page_str[kMaxIntSize];
    snprintf(page_str, kMaxIntSize - 1, "%d", page);
    SetVariable("applybox_page", page_str);
```

At least if they use Tess as a separate process like I do. Which is obviously much easier than using the API and easier to upgrade also.

**Comment by dmigowski** [comment]
This is still open in the release at https://github.com/tesseract-ocr/tesseract/releases/tag/5.5.0
But when I build with vcpkg as instructed, the error doesn't occur anymore. Please fix this. 
Also I do not want all the output unbuffered but just after the tprintf's I located above. And Linux and the vcpkg Build also flush here so it can't be that much of the problem, right?
